### PR TITLE
Issue 17

### PR DESCRIPTION
**Pull request by runcom**

### CODE REVIEW

**Comment by runcom** [comment]
earlier review here https://github.com/runcom/aws-nitro-enclaves-cose/commit/fa811170621589177360312df96d2919a7397fa5

File: src/encrypt.rs [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+
+        let protected_bytes =
+            map_to_empty_or_serialized(&protected).map_err(COSEError::SerializationError)?;
+
+        let enc_structure = EncStructure::new_encrypt0(&protected_bytes)
+            .map_err(COSEError::SerializationError)?;
+
+        let mut tag = vec![0; cose_alg.tag_size()];
+        let mut ciphertext = encrypt_aead(cipher,
+            &key.private_key_to_der().unwrap(),
```
Comment (puiterwijk) [inline src/encrypt.rs]: The `key` here should be the `&[u8]` directly.

File: src/encrypt.rs [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+        let (ciphertext, mut tag) = ciphertext_and_tag.split_at(ciphertext_and_tag.len()-cose_alg.tag_size());
+
+        let payload = decrypt_aead(cipher,
+            &key.private_key_to_der().unwrap(),
+            Some(&iv),
+            &enc_structure
+                .as_bytes()
+                .map_err(COSEError::SerializationError)?,
+            &ciphertext[..],
+            &mut tag).map_err(COSEError::EncryptionError)?;
```
Comment (puiterwijk) [inline src/encrypt.rs]: For `decrypt_aead`, the `tag` should not be mutable.

File: src/header_map.rs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    /// Parses a slice of bytes into a HeaderMap, if possible.
+    pub fn from_bytes(header_map: &[u8]) -> Result<Self, CborError> {
+        serde_cbor::from_slice(&header_map)
+    }
+}
+
+///
+pub fn map_to_empty_or_serialized(map: &HeaderMap) -> Result<Vec<u8>, CborError> {
```
Comment (puiterwijk) [inline src/header_map.rs]: Possibly make this `pub(crate) fn`?

Comment (runcom) [reply src/header_map.rs]: without moving it and the struct to their own file?

File: src/encrypt.rs (hunk 1 above)
Comment (runcom) [reply src/encrypt.rs]: uhm isn't getting the der bytes enough? I've been digging into PKeyRef/PKey and couldn't find anything otherwise

Comment (puiterwijk) [reply src/encrypt.rs]: `PKey` is for `Public/Private keys`, i.e. asymmetric cryptography.

Comment (runcom) [reply src/encrypt.rs]: changed the fns to accept an `&[u8]`

File: src/encrypt.rs [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+///
+pub enum COSEAlgorithm {
+    /// AES-GCM mode w/ 128-bit key, 128-bit tag
+    A128GCM,
+    /// AES-GCM mode w/ 192-bit key, 128-bit tag
+    A192GCM,
+    /// AES-GCM mode w/ 256-bit key, 128-bit tag
+    A256GCM,
+    /// AES-CCM mode 128-bit key, 64-bit tag, 13-byte nonce
+    AESCCM1664128,
```
Comment (puiterwijk) [inline src/encrypt.rs]: This constant is unreadable. Maybe change it to `Aes_Ccm_16_64_128` to also match the capitalizing wishes of clippy?

File: src/encrypt.rs [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+            COSEAlgorithm::AESCCM6464256 => 13,
+            COSEAlgorithm::AESCCM16128128 => 30,
+            COSEAlgorithm::AESCCM16128256 => 31,
+            COSEAlgorithm::AESCCM64128128 => 32,
+            COSEAlgorithm::AESCCM64128256 => 33
+        }
+    }
+
+    ///
+    pub fn to_cose_alg(value: i8) -> Option<COSEAlgorithm> {
```
Comment (puiterwijk) [inline src/encrypt.rs]: Maybe call this `from_value` to match the other function?

File: src/encrypt.rs [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+    AESCCM16128256,
+    /// AES-CCM mode 128-bit key, 128-bit tag, 7-byte nonce
+    AESCCM64128128,
+    /// AES-CCM mode 128-bit key, 128-bit tag, 7-byte nonce
+    AESCCM64128256
+}
+
+impl COSEAlgorithm {
+    ///
+    pub fn value(&self) -> usize {
```
Comment (puiterwijk) [inline src/encrypt.rs]: I think a bunch of functions in this `impl` are public where they don't need to be.
I can't see a reason to have them `pub` unless needed by someone.

File: src/encrypt.rs [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+            key,
+            Some(&iv),
+            &enc_structure
+                .as_bytes()
+                .map_err(COSEError::SerializationError)?,
+            ciphertext,
+            &tag).map_err(COSEError::EncryptionError)
+    }
+
+    /// Serializes the structure for transport / storage. If `tagged` is true, the optional #6.18
```
Comment (puiterwijk) [inline src/encrypt.rs]: `#6.16`

File: src/encrypt.rs [hunk 8]
```diff
@@ -0,0 +1,450 @@
...
+        Ok(COSEEncrypt0{
+            protected: ByteBuf::from(protected_bytes),
+            unprotected,
+            ciphertext: ByteBuf::from(ciphertext)
+        })
+    }
+
+    /// Decrypt the ciphertext in the COSE_Encrypt0 structure.
+    /// https://datatracker.ietf.org/doc/html/rfc8152#section-5.3
+    pub fn decrypt(
```
Comment (puiterwijk) [inline src/encrypt.rs]: I think we might want to return the `protected` and `unprotected` HeaderMap's as well. If not from this one, maybe make an underlying function that returns it all and make this function only return the payload or something?

Comment (runcom) [reply src/encrypt.rs]: changed to

```diff
     pub fn decrypt(
         &self,
         key: &[u8],
-    ) -> Result<Vec<u8>, COSEError> {
+    ) -> Result<(HeaderMap, &HeaderMap, Vec<u8>), COSEError> {
```

**Pull request closed without merging**